### PR TITLE
Reorg fixes

### PR DESCRIPTION
**Pull request by maxwolff**
fixes ToB audit issues 9 and 11 regarding off by one errors in the reorg logic. also adds tests for `formulate_reorg` and `fetch_eth_blocks`

### CODE REVIEW

File: pallets/cash/src/events.rs [hunk 1]
```diff
@@ -70,9 +70,12 @@ fn fetch_eth_blocks(
 ) -> Result<ChainBlocks, EventError> {
     debug!("Fetching Eth Blocks [{}-{}]", from, to);
     let mut acc: Vec<<Ethereum as Chain>::Block> = vec![];
-    for block_number in from..to {
+    for block_number in from..=to {
         match fetch_eth_block(block_number, eth_starport_address) {
             Ok(block) => {
+                // if let Some(last) = acc.last() {
+                //     assert!(last.hash == block.parent_hash, "Fetched blocks do not form chain");
```
Comment (maxwolff) [inline pallets/cash/src/events.rs]: worth it or not?

Comment (jflatow) [reply pallets/cash/src/events.rs]: On the worker side? Good ❓

Comment (waynenilsen) [reply pallets/cash/src/events.rs]: I think the consequences of this check need to be considered in the wider context of how reorgs work. I'm not super familiar with it. Aside from what the knock-on effects are, `assert` would not necessarily be the thing to do, should be `return Err(...)`

File: pallets/cash/src/events.rs [hunk 2]
```diff
@@ -85,3 +88,63 @@ fn fetch_eth_blocks(
...
+
+            match fetched_blocks {
+                ChainBlocks::Eth(blocks) => {
+                    for (expected, actual) in blocks_to_return.iter().zip(blocks.iter()) {
+                        assert_eq!(actual.hash, expected.hash);
+                        assert_eq!(actual.parent_hash, expected.parent_hash);
+                        assert_eq!(actual.number, expected.number);
+                    }
+                }
+                _ => panic!("")
```
Comment (waynenilsen) [inline pallets/cash/src/events.rs]:
```suggestion
                _ => panic!("Expected ChainBlocks::Eth variant")
```

File: pallets/cash/src/internal/events.rs [hunk 3]
```diff
@@ -513,6 +510,84 @@ pub fn receive_chain_reorg<T: Config>(
...
     use crate::tests::*;
+    use sp_core::offchain::testing;
+    use sp_core::offchain::OffchainExt;
+    use ethereum_client::EthereumBlock;
+    use params;
+
+    fn gen_old_blocks(start_block: u64, end_block: u64, start_hash: u8) -> Vec<EthereumBlock> {
+        let mut next_hash = start_hash;
+        let mut v: Vec<ethereum_client::EthereumBlock> = vec![];
+        for i in start_block..=end_block {
```
Comment (waynenilsen) [inline pallets/cash/src/internal/events.rs]: Is `end_block` expected to be inclusive range in the context of this function? I see the callsite and it looks like it is somewhat double counting ? `common_ancestor_idx.clone() - 1`

**Review by waynenilsen — COMMENTED** [review]
looks good overall - some small nits

also - didn't totally grok this change but i didn't look over the ToB report yet so i probably will get it after i read that report 

image

**Comment by jflatow** [comment]
Not done reviewing this myself, and I think 🐺  has a few other tests planned. But I'm rebasing this onto the substrate upgrade / other ToB changes to keep it in sync

File: pallets/cash/src/internal/events.rs (hunk 3 above)
Comment (maxwolff) [reply pallets/cash/src/internal/events.rs]: yes good catch

**Review by hayesgm — COMMENTED** [review]
I'm not sure this has been included in #353, but it will need to be cleaned up for it.

File: pallets/cash/src/internal/events.rs [hunk 4]
```diff
@@ -311,7 +311,7 @@ pub fn formulate_reorg<T: Config>(
             to_hash,
             reverse_blocks: reverse_blocks
                 .into_iter()
-                .take_while(|b| b.parent_hash() != common_ancestor)
+                .take_while(|b| b.hash() != common_ancestor)
```
Comment (maxwolff) [inline pallets/cash/src/internal/events.rs:314]: accidentally didn't include this in my first patch

**Pull request merged**